### ISSUE TITLE
Large delay before advertising load balancer service

### ISSUE THREAD

**Issue by phemmer** [cloudnativelabs/kube-router#426]
There seems to be a very large delay (minutes) before kube-router advertises the address of load balancer services via BGP.

#### 13:47:31 - converted webconsole service from type `ClusterIP` to type `LoadBalancer`:
```
# oc patch service webconsole -p '{"spec":{"type":"LoadBalancer"}}'

# oc get services
NAME         TYPE           CLUSTER-IP     EXTERNAL-IP                 PORT(S)         AGE
webconsole   LoadBalancer   10.128.67.30   10.128.68.95,10.128.68.95   443:30065/TCP   3m
```

The router (bird) logs the following:
```
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosw02ci: Got UPDATE
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosw02ci > ignored 10.128.67.30/32 via 172.17.0.7 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosi01ci: Got UPDATE
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosi01ci > ignored 10.128.67.30/32 via 172.17.0.4 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosm02ci: Got UPDATE
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosm02ci > ignored 10.128.67.30/32 via 172.17.0.3 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosm01ci: Got UPDATE
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosm01ci > ignored 10.128.67.30/32 via 172.17.0.2 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosi02ci: Got UPDATE
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosi02ci > ignored 10.128.67.30/32 via 172.17.0.5 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosw01ci: Got UPDATE
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosw01ci > added 10.128.67.30/32 via 172.17.0.6 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: ...
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosi02ci > ignored 10.128.67.30/32 via 172.17.0.5 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosw02ci > ignored 10.128.67.30/32 via 172.17.0.7 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosi01ci > ignored 10.128.67.30/32 via 172.17.0.4 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosm02ci > ignored 10.128.67.30/32 via 172.17.0.3 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosm01ci > ignored 10.128.67.30/32 via 172.17.0.2 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosw01ci > ignored 10.128.67.30/32 via 172.17.0.6 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosi02ci > ignored 10.128.67.30/32 via 172.17.0.5 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosw02ci > ignored 10.128.67.30/32 via 172.17.0.7 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosm02ci > ignored 10.128.67.30/32 via 172.17.0.3 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosi01ci > ignored 10.128.67.30/32 via 172.17.0.4 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosm01ci > ignored 10.128.67.30/32 via 172.17.0.2 on ens192
May 06 13:47:31 fll2aosdci bird[11686]: fll2aosw01ci > ignored 10.128.67.30/32 via 172.17.0.6 on ens192
```

#### ~13:51:00 - exec into kube-router pod on `fll2aosm01ci` (172.17.0.2)
Unfortunately I don't know the exact time I did this, but it was at approximately 13:51.
```
Here's a quick look at what's happening on this Node
--- BGP Server Configuration ---
AS:        65101
Router-ID: 172.17.0.2
Listening Port: 179, Addresses: 172.17.0.2, ::

--- BGP Neighbors ---
Peer          AS  Up/Down State       |#Received  Accepted
172.17.0.1 65101 00:11:51 Establ      |        1         1
172.17.0.3 65101 00:11:48 Establ      |        1         1
172.17.0.4 65101 00:12:00 Establ      |        1         1
172.17.0.5 65101 00:11:50 Establ      |        1         1
172.17.0.6 65101 00:11:53 Establ      |        1         1
172.17.0.7 65101 00:11:48 Establ      |        1         1

--- BGP Route Info ---
   Network              Next Hop             AS_PATH              Age        Attrs
*> 0.0.0.0/0            10.0.33.1                                 00:11:51   [{Origin: i} {LocalPref: 100}]
*> 10.128.0.0/23        172.17.0.2                                00:02:04   [{Origin: i}]
*> 10.128.2.0/23        172.17.0.3                                00:11:48   [{Origin: i} {LocalPref: 100}]
*> 10.128.4.0/23        172.17.0.4                                00:12:00   [{Origin: i} {LocalPref: 100}]
*> 10.128.6.0/23        172.17.0.5                                00:11:50   [{Origin: i} {LocalPref: 100}]
*> 10.128.8.0/23        172.17.0.6                                00:11:53   [{Origin: i} {LocalPref: 100}]
*> 10.128.10.0/23       172.17.0.7                                00:11:48   [{Origin: i} {LocalPref: 100}]
*> 10.128.64.1/32       172.17.0.2                                00:02:04   [{Origin: i}]
*> 10.128.64.185/32     172.17.0.2                                00:02:04   [{Origin: i}]
*> 10.128.65.92/32      172.17.0.2                                00:02:04   [{Origin: i}]
*> 10.128.65.161/32     172.17.0.2                                00:02:04   [{Origin: i}]
*> 10.128.67.30/32      172.17.0.2                                00:02:04   [{Origin: i}]
*> 10.128.68.95/32      172.17.0.2                                00:02:04   [{Origin: i}]

--- IPVS Services ---
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  172.17.0.2:30065 rr
  -> 10.128.0.3:8443              Masq    1      0          0         
  -> 10.128.2.3:8443              Masq    1      0          0         
TCP  10.128.64.1:53 rr persistent 10800
  -> 172.17.0.2:8053              Masq    1      0          0         
  -> 172.17.0.3:8053              Masq    1      0          0         
TCP  10.128.64.1:443 rr persistent 10800
  -> 172.17.0.2:8443              Masq    1      0          0         
  -> 172.17.0.3:8443              Masq    1      0          0         
TCP  10.128.64.185:9000 rr
  -> 10.128.4.4:9090              Masq    1      0          0         
TCP  10.128.65.92:5000 rr persistent 10800
  -> 10.128.4.3:5000              Masq    1      0          0         
TCP  10.128.65.161:80 rr
  -> 172.17.0.4:80                Masq    1      0          0         
  -> 172.17.0.5:80                Masq    1      0          0         
TCP  10.128.65.161:443 rr
  -> 172.17.0.4:443               Masq    1      0          0         
  -> 172.17.0.5:443               Masq    1      0          0         
TCP  10.128.65.161:1936 rr
  -> 172.17.0.4:1936              Masq    1      0          0         
  -> 172.17.0.5:1936              Masq    1      0          0         
TCP  10.128.67.30:443 rr
  -> 10.128.0.3:8443              Masq    1      0          0         
  -> 10.128.2.3:8443              Masq    1      0          0         
TCP  10.128.68.95:443 rr
  -> 10.128.0.3:8443              Masq    1      0          0         
  -> 10.128.2.3:8443              Masq    1      0          0         
UDP  10.128.64.1:53 rr persistent 10800
  -> 172.17.0.2:8053              Masq    1      0          0         
  -> 172.17.0.3:8053              Masq    1      0          0         
```

^ as you can see, it claims that `10.128.68.95` is being advertised via BGP. However it is not.

#### 13:55:43 - router finally receives advertisement for `10.128.68.95`:
```
May 06 13:55:43 fll2aosdci bird[11686]: fll2aosi01ci > added [best] 10.128.68.95/32 via 172.17.0.4 on ens192
May 06 13:55:43 fll2aosdci bird[11686]: kernel1 < added 10.128.68.95/32 via 172.17.0.4 on ens192
May 06 13:55:43 fll2aosdci bird[11686]: fll2aosi01ci < rejected by protocol 10.128.68.95/32 via 172.17.0.4 on ens192
May 06 13:55:43 fll2aosdci bird[11686]: fll2aosi02ci < rejected by protocol 10.128.68.95/32 via 172.17.0.4 on ens192
May 06 13:55:43 fll2aosdci bird[11686]: fll2aosw02ci < rejected by protocol 10.128.68.95/32 via 172.17.0.4 on ens192
May 06 13:55:43 fll2aosdci bird[11686]: fll2aosm02ci < rejected by protocol 10.128.68.95/32 via 172.17.0.4 on ens192
May 06 13:55:43 fll2aosdci bird[11686]: fll2aosm01ci < rejected by protocol 10.128.68.95/32 via 172.17.0.4 on ens192
May 06 13:55:43 fll2aosdci bird[11686]: fll2aosw01ci < rejected by protocol 10.128.68.95/32 via 172.17.0.4 on ens192
May 06 13:55:44 fll2aosdci bird[11686]: kernel1: 10.128.68.95/32: seen
May 06 13:55:46 fll2aosdci bird[11686]: kernel1: 10.128.68.95/32: seen
May 06 13:55:47 fll2aosdci bird[11686]: fll2aosw01ci > added 10.128.68.95/32 via 172.17.0.6 on ens192
May 06 13:55:48 fll2aosdci bird[11686]: kernel1: 10.128.68.95/32: seen
May 06 13:55:48 fll2aosdci bird[11686]: fll2aosi02ci > added 10.128.68.95/32 via 172.17.0.5 on ens192
May 06 13:55:48 fll2aosdci bird[11686]: fll2aosw02ci > added 10.128.68.95/32 via 172.17.0.7 on ens192
May 06 13:55:48 fll2aosdci bird[11686]: fll2aosm02ci > added [best] 10.128.68.9/32 via 172.17.0.3 on ens192
May 06 13:55:48 fll2aosdci bird[11686]: kernel1 < replaced 10.128.68.95/32 via 172.17.0.3 on ens192
May 06 13:55:48 fll2aosdci bird[11686]: fll2aosi01ci < rejected by protocol 10.128.68.95/32 via 172.17.0.3 on ens192
May 06 13:55:48 fll2aosdci bird[11686]: fll2aosi02ci < rejected by protocol 10.128.68.95/32 via 172.17.0.3 on ens192
May 06 13:55:48 fll2aosdci bird[11686]: fll2aosw02ci < rejected by protocol 10.128.68.95/32 via 172.17.0.3 on ens192
May 06 13:55:48 fll2aosdci bird[11686]: fll2aosm02ci < rejected by protocol 10.128.68.95/32 via 172.17.0.3 on ens192
May 06 13:55:48 fll2aosdci bird[11686]: fll2aosm01ci < rejected by protocol 10.128.68.95/32 via 172.17.0.3 on ens192
May 06 13:55:48 fll2aosdci bird[11686]: fll2aosw01ci < rejected by protocol 10.128.68.95/32 via 172.17.0.3 on ens192
May 06 13:55:49 fll2aosdci bird[11686]: fll2aosm01ci > added [best] 10.128.68.9/32 via 172.17.0.2 on ens192
May 06 13:55:49 fll2aosdci bird[11686]: kernel1 < replaced 10.128.68.95/32 via 172.17.0.2 on ens192
May 06 13:55:49 fll2aosdci bird[11686]: fll2aosi01ci < rejected by protocol 10.128.68.95/32 via 172.17.0.2 on ens192
May 06 13:55:49 fll2aosdci bird[11686]: fll2aosi02ci < rejected by protocol 10.128.68.95/32 via 172.17.0.2 on ens192
May 06 13:55:49 fll2aosdci bird[11686]: fll2aosw02ci < rejected by protocol 10.128.68.95/32 via 172.17.0.2 on ens192
May 06 13:55:49 fll2aosdci bird[11686]: fll2aosm02ci < rejected by protocol 10.128.68.95/32 via 172.17.0.2 on ens192
May 06 13:55:49 fll2aosdci bird[11686]: fll2aosm01ci < rejected by protocol 10.128.68.95/32 via 172.17.0.2 on ens192
May 06 13:55:49 fll2aosdci bird[11686]: fll2aosw01ci < rejected by protocol 10.128.68.95/32 via 172.17.0.2 on ens192
May 06 13:55:50 fll2aosdci bird[11686]: kernel1: 10.128.68.95/32: seen
May 06 13:55:52 fll2aosdci bird[11686]: kernel1: 10.128.68.95/32: seen
May 06 13:55:54 fll2aosdci bird[11686]: kernel1: 10.128.68.95/32: seen
May 06 13:55:56 fll2aosdci bird[11686]: kernel1: 10.128.68.95/32: seen
```
^ also notice it flops around a bit until it finally settles on 172.17.0.2

This is with:
OpenShift 3.9.0
Kubernetes 1.9.1
kube-router docker.io/cloudnativelabs/kube-router@sha256:a6a30873d1676beba0d79fe5bd9ae68ad0c28cf11785475e2454b736d1e6ecf1

kube-router args:
```
    - --run-router=true
    - --run-firewall=true
    - --run-service-proxy=true
    - --kubeconfig=/var/lib/kube-router/kubeconfig
    - --bgp-graceful-restart=true
    - --cluster-cidr=10.128.0.0/18
    - --cluster-asn=65101
    - --peer-router-ips=172.17.0.1
    - --peer-router-asns=65101
    - --peer-router-passwords=
    - --enable-overlay=false
    - --enable-pod-egress=false
    - --advertise-cluster-ip=true
    - --advertise-external-ip=true
    - --advertise-loadbalancer-ip=true
    - --hairpin-mode
```

**Comment by murali-reddy** [comment]
@phemmer thanks for detailed observations. Earlier there used to be periodic sync with duration of 1 min which was bumped to 5 min. These periodic sync's are just meant to be reconciliation of desired state. But on any relevent events (like service add/delete, endpoint add/delete etc) kube-router acts on the event and perform the sync's. 

From your observation, when full sync happens (ie. at 5 min duration) things are workign as expected but not in-line with events of service add/delete etc. I will take a look where the missing functionality is.

**Comment by phemmer** [comment]
I bumped the logging level up and have the logs from one of the kube-router pods below. According to the logs it's receiving the watch notification, and claims to be advertising it, so dunno.

Logs filtered for `10.128.67.80` or `sync` or `test`
```
I0512 20:47:46.835258       1 network_services_controller.go:395] Received update to service: default/test from watch API
I0512 20:47:46.835531       1 network_services_controller.go:410] Syncing IPVS services sync on update to service: default/test
I0512 20:47:46.835446       1 ecmp_vip.go:84] Received update to service: default/test from watch API
I0512 20:47:46.836808       1 ecmp_vip.go:24] Advertising route: '10.128.67.80/32 via 172.17.0.2' to peers
I0512 20:47:46.843387       1 network_services_controller.go:1381] Successfully added service: TCP:10.128.67.80:9000 (Flags: )
I0512 20:47:46.844480       1 network_services_controller.go:761] IPVS servers and services are synced to desired state
I0512 20:47:46.844559       1 network_services_controller.go:438] sync ipvs services took 8.871832ms
I0512 20:47:46.862256       1 network_services_controller.go:365] Received update to endpoint: default/test from watch API
I0512 20:47:46.862420       1 network_services_controller.go:380] Syncing IPVS services sync for update to endpoint: default/test
I0512 20:47:46.863314       1 ecmp_vip.go:161] Received update to endpoint: default/test from watch API
I0512 20:47:46.863438       1 ecmp_vip.go:24] Advertising route: '10.128.67.80/32 via 172.17.0.2' to peers
I0512 20:47:46.870580       1 network_services_controller.go:1478] Successfully added destination 10.128.4.4:9090 (Weight: 1) to the service TCP:10.128.67.80:9000 (Flags: [hashed entry])
I0512 20:47:46.872168       1 network_services_controller.go:761] IPVS servers and services are synced to desired state
I0512 20:47:46.872238       1 network_services_controller.go:438] sync ipvs services took 9.69728ms
I0512 20:49:44.998579       1 network_policy_controller.go:136] Performing periodic sync of iptables to reflect network policies
I0512 20:49:44.998861       1 network_policy_controller.go:216] Starting periodic sync of iptables
I0512 20:49:45.002813       1 network_policy_controller.go:324] Iptables chains in the filter table are synchronized with the network policies.
I0512 20:49:45.013561       1 network_policy_controller.go:213] sync iptables took 14.702291ms
I0512 20:49:45.028482       1 network_services_controller.go:244] Performing periodic sync of ipvs services
I0512 20:49:45.113929       1 network_services_controller.go:761] IPVS servers and services are synced to desired state
I0512 20:49:45.114055       1 network_services_controller.go:438] sync ipvs services took 11.085922ms
I0512 20:49:46.090967       1 network_routes_controller.go:253] Performing periodic sync of service VIP routes
I0512 20:49:46.094961       1 ecmp_vip.go:24] Advertising route: '10.128.67.80/32 via 172.17.0.2' to peers
I0512 20:49:46.095873       1 network_routes_controller.go:257] Performing periodic sync of pod CIDR routes
I0512 20:54:44.998459       1 network_policy_controller.go:136] Performing periodic sync of iptables to reflect network policies
I0512 20:54:44.998688       1 network_policy_controller.go:216] Starting periodic sync of iptables
I0512 20:54:45.001559       1 network_policy_controller.go:324] Iptables chains in the filter table are synchronized with the network policies.
I0512 20:54:45.010882       1 network_policy_controller.go:213] sync iptables took 12.190126ms
I0512 20:54:45.028623       1 network_services_controller.go:244] Performing periodic sync of ipvs services
I0512 20:54:45.121876       1 network_services_controller.go:761] IPVS servers and services are synced to desired state
I0512 20:54:45.121963       1 network_services_controller.go:438] sync ipvs services took 11.43729ms
I0512 20:54:46.091016       1 network_routes_controller.go:253] Performing periodic sync of service VIP routes
I0512 20:54:46.093568       1 ecmp_vip.go:24] Advertising route: '10.128.67.80/32 via 172.17.0.2' to peers
I0512 20:54:46.095332       1 network_routes_controller.go:257] Performing periodic sync of pod CIDR routes
```

* 20:47:46 - service created
* 20:49:45 - IP shows up in `gobgp neighbor 172.17.0.1 adj-out`
* 20:54:45 - route shows up on router

I've repeated this many more times, and on rare occasion the IP will show up in `gobgp neighbor ... adj-out` and on the router at the same time. Maybe like a 1 in 6 chance. But even when that happens, it's still not until one of the sync intervals occurs.

I've done a packet capture on the router to make sure the router isn't the problem, or to see if the routes are coming in mangled somehow, but nothing comes in at all. Capture is silent except for the occasional keepalive.

**Comment by murali-reddy** [comment]
@phemmer ok I figured whats happening. So kube-router uses export policies to control what is advertised. So when service is added/deleted RIB is updated but corresponding export policies are not updated. 

You shall see a prefix set like below, which should get updated with service VIP's.
```
root@kube-node1:~ #gobgp policy prefix clusteripprefixset
NAME                PREFIX
clusteripprefixset  10.96.0.1/32 32..32
                    10.96.0.10/32 32..32
                    10.106.112.219/32 32..32
                    10.107.52.30/32 32..32
                    10.111.189.86/32 32..32
```
`clusteripprefixset` is getting updated only during periodic sync which is 5 min duration by default.

I have opened https://github.com/cloudnativelabs/kube-router/pull/435

Also if you want to test please try image `cloudnativelabs/kube-router-git:issue-426`

**Comment by phemmer** [comment]
Thanks! That does appear to have fixed the issue.
The first time I created a new service with this fixed version, the router only received routes for half the nodes. The other half didn't show up until the next sync (and the missing nodes did have it listed in their policies). But this only happened the first time. I've done it about 10 more times and it hasn't happened again.

**Comment by phemmer** [comment]
Actually, nix that, I can reproduce it. It seems to happen the first time after kube-router is started.

**Comment by phemmer** [comment]
I think I figured it out. It's more graceful restart issues.
I noticed something in the router's logs as soon as the missing nodes started working:
```
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: Incoming connection from 172.17.0.5 (port 48056) accepted
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: Sending OPEN(ver=4,as=65101,hold=240,id=0a0021fc)
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: Connecting to 172.17.0.5 from local address 172.17.0.1
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: Got OPEN(as=65101,hold=90,id=ac110005)
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: BGP session established
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: State changed to feed
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < added 0.0.0.0/0 via 10.0.33.1 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.10.0/23 via 172.17.0.7 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.8.0/23 via 172.17.0.6 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.2.0/23 via 172.17.0.3 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.0.0/23 via 172.17.0.2 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.6.0/23 via 172.17.0.5 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.4.0/23 via 172.17.0.4 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.67.22/32 via 172.17.0.2 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.66.213/32 via 172.17.0.2 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.64.149/32 via 172.17.0.2 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.67.220/32 via 172.17.0.2 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.64.1/32 via 172.17.0.2 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.66.134/32 via 172.17.0.2 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.65.125/32 via 172.17.0.2 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.64.230/32 via 172.17.0.2 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.67.165/32 via 172.17.0.2 on ens192
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: State changed to up
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: Sending UPDATE
May 12 19:36:03 fll2aosdci bird[11656]: fll2aosi02ci: Sending End-of-RIB
May 12 19:36:30 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:36:33 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:36:42 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.66.89/32 via 172.17.0.3 on ens192
May 12 19:36:59 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:37:03 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:37:25 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:37:33 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:37:54 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:38:03 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:38:21 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:38:33 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:38:49 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:39:03 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:39:15 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:39:33 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:39:38 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:40:03 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:40:03 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:40:31 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:40:33 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:41:01 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:41:03 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:41:25 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:41:33 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:41:49 fll2aosdci bird[11656]: fll2aosi02ci: Sending KEEPALIVE
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.67.220/32 via 172.17.0.2 on ens192
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.65.125/32 via 172.17.0.2 on ens192
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.0.0/23 via 172.17.0.2 on ens192
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.64.230/32 via 172.17.0.2 on ens192
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.66.134/32 via 172.17.0.2 on ens192
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.64.1/32 via 172.17.0.2 on ens192
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.67.22/32 via 172.17.0.2 on ens192
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.66.89/32 via 172.17.0.2 on ens192
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.66.213/32 via 172.17.0.4 on ens192
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.64.149/32 via 172.17.0.3 on ens192
May 12 19:41:58 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.67.165/32 via 172.17.0.4 on ens192
May 12 19:41:59 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.8.0/23 via 172.17.0.6 on ens192
May 12 19:42:01 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.10.0/23 via 172.17.0.7 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci: Got KEEPALIVE
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci: Got UPDATE
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > added 10.128.65.125/32 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > added 10.128.66.89/32 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > added [best] 10.128.6.0/23 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci < rejected by protocol 10.128.6.0/23 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > added 10.128.67.220/32 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > added 10.128.66.134/32 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > added 10.128.64.230/32 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > added 10.128.64.1/32 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > added 10.128.67.22/32 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci: Got UPDATE
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci: Got End-of-RIB
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci: Neighbor graceful restart done
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > removed 10.128.66.213/32 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > removed 10.128.64.149/32 via 172.17.0.5 on ens192
May 12 19:42:03 fll2aosdci bird[11656]: fll2aosi02ci > removed 10.128.67.165/32 via 172.17.0.5 on ens192
```

Note that it connected at 19:36:03, but `Neighbor graceful restart done` didn't come in until 19:42:03, which is when the missing service IP came in.

Now as for why the nodes apparently have a 50/50 chance of getting into this state, I don't know. But I think that's a different issue entirely (and will continue poking into it). So this issue can be marked as fully resolved by #435

**Comment by debianmaster** [comment]
@phemmer  do you mind sharing a README of how you configured kube-router to openshift?
